Wrong Action may be executed after detaching a component

hesara opened the pull request:
Due to action key generation on ActionManager a wrong action
may be executed if the component that fired the wanted action
is already detached.

This patch makes action keys globally unique, simplifying the
approach of #8495 but reusing its tests.

Fixes #5864

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/9806)
<!-- Reviewable:end -->

mcollovati committed:
Wrong Action may be executed when for detached component (#5864)

Due to action key generation on ActionManager a wrong action
may be executed if the component that fired the wanted action
is already detached.
Change-Id: Ib0a0bfac2e11d8e2aa0d4fbf1f9eff3383c488d6

Commit:
Simplify code and fix tests

The action manager now uses globally unique keys.

Ideally the same action should keep its key across requests.
The original implementation only did this by chance when the
set of actions did not change, and the current implementation
does not try to preserve keys as they are resent on every
paintActions() call.

Commit:
Minor fixes and javadoc

Commit:
Fix WindowAndUIShortcutsTest

tsuoanttila commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> on my part after the few sinces have been filled.

---

Reviewed 7 of 8 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[server/src/main/java/com/vaadin/event/ActionManager.java, line 265 at r2](https://reviewable.io:443/reviews/vaadin/framework/9806#-KrAkKh75nHCmiuhIHsj:-KrAkKh75nHCmiuhIHsk:b-oagfb2) ([raw file](https://github.com/vaadin/framework/blob/f0a475754dce278fe24ac9908abfd321a13a9705/server/src/main/java/com/vaadin/event/ActionManager.java#L265)):*
> ```Java
>      * instances.
>      *
>      * @since
> ```

empty since

---

*[server/src/main/java/com/vaadin/server/KeyMapper.java, line 98 at r2](https://reviewable.io:443/reviews/vaadin/framework/9806#-KrAkQrJtIqPuSAInsG1:-KrAkQrJtIqPuSAInsG2:b-oagfb2) ([raw file](https://github.com/vaadin/framework/blob/f0a475754dce278fe24ac9908abfd321a13a9705/server/src/main/java/com/vaadin/server/KeyMapper.java#L98)):*
> ```Java
>      *
>      * @return new key
>      * @since
> ```

empty since

---

*[uitest/src/test/java/com/vaadin/tests/components/ui/WindowAndUIShortcutsTest.java, line 38 at r2](https://reviewable.io:443/reviews/vaadin/framework/9806#-KrAkpd1gmTdbi8ivBVb:-KrAkpd1gmTdbi8ivBVc:bulhx2h) ([raw file](https://github.com/vaadin/framework/blob/f0a475754dce278fe24ac9908abfd321a13a9705/uitest/src/test/java/com/vaadin/tests/components/ui/WindowAndUIShortcutsTest.java#L38)):*
> ```Java
> 
>         WindowElement window = $(WindowElement.class).first();
>         // for PhantomJS to have the focus in the right place
> ```

We'll hopefully stop testing with PhantomJS soon as the headless versions of other browsers are around the corner.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9806#-:-KrAl6WLS5l0SnwZhc0S:bxd5h2n)*
<!-- Sent from Reviewable.io -->

Commit:
Fill in since tags

hesara commented:
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/com/vaadin/event/ActionManager.java, line 265 at r2](https://reviewable.io:443/reviews/vaadin/framework/9806#-KrAkKh75nHCmiuhIHsj:-KrAmUHipR64IY4Q0EIi:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/f0a475754dce278fe24ac9908abfd321a13a9705/server/src/main/java/com/vaadin/event/ActionManager.java#L265)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

empty since
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/server/KeyMapper.java, line 98 at r2](https://reviewable.io:443/reviews/vaadin/framework/9806#-KrAkQrJtIqPuSAInsG1:-KrAmUmLj3Z4fOy8D8vI:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/f0a475754dce278fe24ac9908abfd321a13a9705/server/src/main/java/com/vaadin/server/KeyMapper.java#L98)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

empty since
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9806)*
<!-- Sent from Reviewable.io -->

Commit:
Use AtomicInteger as static counter

This provides thread safety when multiple sessions share the
ClassLoader used for ActionKeyMapper.

tsuoanttila commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9806#-:-KrFAeehwbKDwTD49yJH:bnfp4nl)*
<!-- Sent from Reviewable.io -->
